### PR TITLE
PP-4375 Stripe source notifications

### PR DESCRIPTION
**Pull request by kbottla**
## WHAT
Processes the following source notification types from Stripe (for charges in  AUTHORISATION_3DS_REQUIRED state)

| Notification type |  updates charge status to     |
|-------------------|--------------------------|
| source.chargeable | AUTHORISATION_3DS_READY  |
| source.failed     | AUTHORISATION_REJECTED   |
| source.canceled   | AUTHORISATION_CANCELLED  |

Any other notification is not handled

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/gateway/stripe/StripeNotification.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
+package uk.gov.pay.connector.gateway.stripe;
```
Comment (oswaldquek) [inline src/main/java/uk/gov/pay/connector/gateway/stripe/StripeNotification.java]: can this be in package `uk.gov.pay.connector.gateway.stripe.json` please

File: src/main/java/uk/gov/pay/connector/gateway/stripe/StripeNotification.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.JsonNode;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class StripeNotification {
+
+    @JsonProperty("id")
+    String id;
```
Comment (oswaldquek) [inline src/main/java/uk/gov/pay/connector/gateway/stripe/StripeNotification.java]: fields should be private

File: src/main/java/uk/gov/pay/connector/gateway/stripe/StripeNotification.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    @JsonProperty("type")
+    String type;
```
Comment (oswaldquek) [inline src/main/java/uk/gov/pay/connector/gateway/stripe/StripeNotification.java]: fields should be private

File: src/main/java/uk/gov/pay/connector/gateway/stripe/StripeNotification.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    @JsonProperty("data")
+    StripeEventData data;
```
Comment (oswaldquek) [inline src/main/java/uk/gov/pay/connector/gateway/stripe/StripeNotification.java]: fields should be private

File: src/main/java/uk/gov/pay/connector/gateway/stripe/StripeNotificationType.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import java.util.Arrays;
+
+public enum StripeNotificationType {
+
+    SOURCE_CANCELED("source.canceled"),
+    SOURCE_CHARGEABLE("source.chargeable"),
+    SOURCE_FAILED("source.failed"),
+    UNKNOWN("");
+
+    public final String type;
```
Comment (oswaldquek) [inline src/main/java/uk/gov/pay/connector/gateway/stripe/StripeNotificationType.java]: I don't mind this being `public final` but it makes the `getType()` pointless. I guess either 1) make this private or 2) delete `getType()` method.

File: src/test/java/uk/gov/pay/connector/it/resources/stripe/StripeNotificationResourceITest.java [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+import static uk.gov.pay.connector.gateway.PaymentGatewayName.STRIPE;
+import static uk.gov.pay.connector.gateway.stripe.StripeNotificationType.SOURCE_CANCELED;
+import static uk.gov.pay.connector.gateway.stripe.StripeNotificationType.SOURCE_CHARGEABLE;
+import static uk.gov.pay.connector.gateway.stripe.StripeNotificationType.SOURCE_FAILED;
+import static uk.gov.pay.connector.util.TestTemplateResourceLoader.STRIPE_NOTIFICATION_3DS_SOURCE;
+import static uk.gov.pay.connector.util.TransactionId.randomId;
+
+@RunWith(DropwizardJUnitRunner.class)
+@DropwizardConfig(app = ConnectorApp.class, config = "config/test-it-config.yaml")
+public class StripeNotificationResourceITest extends ChargingITestBase {
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/it/resources/stripe/StripeNotificationResourceITest.java]: Can this test not extend ChargingITestBase please, as ChargingITestBase contains whole lot of stuff that StripeNotificationResourceITest doesn't need. This can only lead to confusion later on. If you have a look at StripeResourceAuthorizeITest, it's really not that bad in terms of needing various fields and setting up.

File: src/test/java/uk/gov/pay/connector/it/resources/stripe/StripeNotificationResourceITest.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+                .extract().body()
+                .asString();
+
+        assertThat(response, is(RESPONSE_EXPECTED_BY_STRIPE));
+
+        assertFrontendChargeStatusIs(externalChargeId, AUTHORISATION_CANCELLED.getValue());
+    }
+
+    @Test
+    public void shouldFailAStripeNotificationWithAnUnexpectedContentType() throws Exception {
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/it/resources/stripe/StripeNotificationResourceITest.java]: unneeded `throws Exception`

File: src/test/java/uk/gov/pay/connector/gateway/stripe/StripeNotificationServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    public void setup() {
+        notificationService = new StripeNotificationService(mockChargeDao,
+                mockChargeNotificationProcessor);
+
+        when(mockCharge.getStatus()).thenReturn(AUTHORISATION_3DS_REQUIRED.getValue());
+        when(mockChargeDao.findByProviderAndTransactionId(STRIPE.getName(), sourceId)).thenReturn(Optional.of(mockCharge));
+    }
+
+    @Test
+    public void shouldUpdateCharge_WhenNotificationIsFor3DSSourceChargeable() {
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/gateway/stripe/StripeNotificationServiceTest.java:57]: what's the value of this test if it's already covered by StripeNotificationResourceITest.shouldHandleASourceChargeableNotification?

File: src/test/java/uk/gov/pay/connector/gateway/stripe/StripeNotificationServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        final String payload = sampleStripeNotification(STRIPE_NOTIFICATION_3DS_SOURCE,
+                sourceId, SOURCE_CHARGEABLE);
+        notificationService.handleNotificationFor(payload);
+
+        verify(mockChargeNotificationProcessor).invoke(sourceId, mockCharge, AUTHORISATION_3DS_READY, null);
+    }
+
+    @Test
+    public void shouldUpdateCharge_WhenNotificationIsFor3DSSourceFailed() {
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/gateway/stripe/StripeNotificationServiceTest.java:66]: what's the value of this test if it's already covered by StripeNotificationResourceITest.shouldHandleASourceFailedNotification?

File: src/test/java/uk/gov/pay/connector/gateway/stripe/StripeNotificationServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        final String payload = sampleStripeNotification(STRIPE_NOTIFICATION_3DS_SOURCE,
+                sourceId, SOURCE_FAILED);
+        notificationService.handleNotificationFor(payload);
+
+        verify(mockChargeNotificationProcessor).invoke(sourceId, mockCharge, AUTHORISATION_REJECTED, null);
+    }
+
+    @Test
+    public void shouldUpdateCharge_WhenNotificationIsFor3DSSourceCancelled() {
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/gateway/stripe/StripeNotificationServiceTest.java:75]: what's the value of this test if it's already covered by StripeNotificationResourceITest.shouldHandleASourceCanceledNotification?

**Review by oswaldquek — CHANGES_REQUESTED** [review]
Nice work. Just a few comments that shouldn't take long to address.

File: src/main/java/uk/gov/pay/connector/gateway/stripe/StripeNotification.java (hunk 1 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/gateway/stripe/StripeNotification.java]: 👍

File: src/main/java/uk/gov/pay/connector/gateway/stripe/StripeNotification.java (hunk 2 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/gateway/stripe/StripeNotification.java]: oops, missed it. Updated to private

File: src/main/java/uk/gov/pay/connector/gateway/stripe/StripeNotification.java (hunk 3 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/gateway/stripe/StripeNotification.java]: Updated to private

File: src/main/java/uk/gov/pay/connector/gateway/stripe/StripeNotification.java (hunk 4 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/gateway/stripe/StripeNotification.java]: Updated to private

File: src/main/java/uk/gov/pay/connector/gateway/stripe/StripeNotificationType.java (hunk 5 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/gateway/stripe/StripeNotificationType.java]: 👍 Updated to private

File: src/test/java/uk/gov/pay/connector/gateway/stripe/StripeNotificationServiceTest.java (hunk 8 above)
Comment (kbottla) [reply src/test/java/uk/gov/pay/connector/gateway/stripe/StripeNotificationServiceTest.java:57]: I feel it is better to have all possibilities/edge cases covered in unit tests. Integration tests shouldn't be exhaustive and there could possibly be few that can overlap. In this all 3 states which doesn't seem to be overkill.

File: src/test/java/uk/gov/pay/connector/gateway/stripe/StripeNotificationServiceTest.java (hunk 9 above)
Comment (kbottla) [reply src/test/java/uk/gov/pay/connector/gateway/stripe/StripeNotificationServiceTest.java:66]: same as above

File: src/test/java/uk/gov/pay/connector/gateway/stripe/StripeNotificationServiceTest.java (hunk 10 above)
Comment (kbottla) [reply src/test/java/uk/gov/pay/connector/gateway/stripe/StripeNotificationServiceTest.java:75]: same as above

File: src/test/java/uk/gov/pay/connector/it/resources/stripe/StripeNotificationResourceITest.java (hunk 7 above)
Comment (kbottla) [reply src/test/java/uk/gov/pay/connector/it/resources/stripe/StripeNotificationResourceITest.java]: 👍

File: src/test/java/uk/gov/pay/connector/it/resources/stripe/StripeNotificationResourceITest.java (hunk 6 above)
Comment (kbottla) [reply src/test/java/uk/gov/pay/connector/it/resources/stripe/StripeNotificationResourceITest.java]: 👍

**Review by oswaldquek — APPROVED** [review]
